indica/converters/time.py

```diff
@@ -186,20 +287,165 @@ def bin_in_time(
         Array like the input, but binned along the time axis.
 
     """
-    npoints = round(abs(tend - tstart) * frequency) + 1
-    half_interval = 0.5 * (tend - tstart) / (npoints - 1)
+    check_bounds_bin(tstart, tend, 1.0 / frequency, data)
+    tlabels = get_tlabels(tstart, tend, frequency)
+
+    return bin_to_time_labels(tlabels, data)
+
+
+def bin_in_time_dt(tstart: float, tend: float, dt: float, data: DataArray) -> DataArray:
+    """Bin given data along the time axis, discarding data before or after
+    the limits.
+
+    Parameters
+    ----------
+    tstart
+        The lower limit in time for determining which data to retain.
+    tend
+        The upper limit in time for determining which data to retain.
+    dt
+        Time resolution of new time axis.
+    data
+        Data to be binned.
+
+    Returns
+    -------
+    :
+        Array like the input, but binned along the time axis.
+
+    """
+    check_bounds_bin(tstart, tend, dt, data)
+    tlabels = get_tlabels_dt(tstart, tend, dt)
+    return bin_to_time_labels(tlabels, data)
+
+
+def get_tlabels(tstart: float, tend: float, frequency: float):
+    """
+    Build time array given start, end and frequency
+
+    Parameters
+    ----------
+    tstart
+        The lower limit in time for determining which data to retain.
+    tend
+        The upper limit in time for determining which data to retain.
+    frequency
+        Frequency of sampling on the time axis.
+
+    Returns
+    -------
+    tlabels
+        Time array
+
+    """
+    npoints = round((tend - tstart) * frequency) + 1
+    return np.linspace(tstart, tend, npoints)
+
+
+def get_tlabels_dt(tstart: float, tend: float, dt: float):
+    """
+    Build time array given start, end and frequency
+
+    Parameters
+    ----------
+    tstart
+        The lower limit in time for determining which data to retain.
+    tend
+        The upper limit in time for determining which data to retain.
+    dt
+        Time resolution of new time axis.
+
+    Returns
+    -------
+    tlabels
+        Time array
+
+    """
+    tlabels = np.arange(tstart, tend + 2 * dt, dt)
+    tlabels = tlabels[np.where((tlabels >= tstart) * (tlabels <= tend))[0]]
+    return tlabels
+
+
+def check_bounds_bin(tstart: float, tend: float, dt: float, data: DataArray):
+    """
+    Check necessary bounds for binning data in time
+
+    Parameters
+    ----------
+    tstart
+        The lower limit in time for determining which data to retain.
+    tend
+        The upper limit in time for determining which data to retain.
+    dt
+        Time resolution of new time axis.
+    data
+        Data to be binned.
+    """
     tcoords = data.coords["t"]
+    half_interval = dt / 2
     if tcoords[0] > tstart + half_interval:
-        raise ValueError(
+        return ValueError(
```

Comment:
Should still be raising the `ValueError` rather than returning, that's failing a couple of tests